docs/rfc/rfc-026-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,375 @@
+A peer should not be banned the first time this happens, but rather if this happens a predefined number 
+of times (`numFailures`) within a time interval (`lastFailure`). This time interval should be reset every `failureResetInterval`ms.  
+
+For each peer, we should have a separate `numFailures` and `lastFailure` variable. There is no need to  have one per transaction. Whenever a transaction fails, if the `now - lastFailure <= failureResetInterval`, then we increment the `numFailures` for this particular peer and set the `lastFailure` to `now`. Otherwise, we set `lastFailure` to `now` but do not increment the `numFailures`  variable. Once the value for `numFailures` for a peer reaches `maxAllowedFailures`, the peer is disconnected from and banned. 
+
+The reason for this logic is as follows: We deem it acceptable if every now and then a peer sends us an invalid transaction. But if this happens very frequently, then this behaviour can be considered as spamming and we want to disconnect from the peer. The actual values of these parameter are yet to be defined and should be based on the time it usually takes for a transacation to travel between two nodes, the time for a block to be agreed on and committed. [ToDo] What other factors can impact their value?
+
+*Banning a peer in case of duplicate transactions*
+
+Currently, a peer can send the same valid (or invalid) transaction [multiple times](https://github.com/tendermint/tendermint/blob/ff0f98892f24aac11e46aeff2b6d2c0ad816701a/mempool/v0/clist_mempool.go#L247). Peers do not 
+gossip transactions to peers that have sent them that same transaction. But there is no check on whether 
+a node has alrady sent the same transaction to this peer before. There is also no check whether the transaction
+that is being gossiped was valid or not (assumint that invalid transactions could become valid).
+The transaction broadcast logic simply loops through the mempool and tries to send the transactions currently in the pool. 
+
+If we want to ban peers based on duplicate transactions, we should either add additional checks for the cases above, or 
+not ban peers for this behaviour at the moment. It would be useful to gather metrics on how often a peer gossips the same 
+transaction and whether this is cause of significant traffic. 
+
+#### **Banning in case of ResponseCheckTx.code > 1**
+
+If a transaction fails since it could never have been valid, `CheckTx` returns a `ResponseCheckTx.code` 
+value greater than 1. In this case, the peer should be disconnected from and banned immediately without keeping count on how often 
+this has happened. 
+
+The question is whether this transaction should be kept track of in the cache? We can still store it in 
+the cache so that we don't run `CheckTx` on it again, but if this peer is immediately banned, maybe there is no need
+to store its information.
```

Comment:
I would say that, if we banned a peer due to a single transaction, we need to keep this info somehow, in case other peers send that TX again. IIUR, the current behavior would just be to "drop" the TX as it is in the cache.